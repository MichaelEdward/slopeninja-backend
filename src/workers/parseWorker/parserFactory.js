import cheerio from 'cheerio';
import { AllHtmlEntities } from 'html-entities';

const htmlEntities = new AllHtmlEntities();

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const xyz = (text) => {
  const decoded = text ? text.toString().replace(/\u0026amp;/gu, '__JAKE__') : null;
  return decoded;
}

export const decodeEntities = (text) => {
  // best way I found to normalize unicode in input
  const unicodeInput = JSON.stringify(JSON.parse(text.toString()));
  return htmlEntities.decode(unicodeInput);
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html), { decodeEntities: true });

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser, preParser = t => t) => async (data = "{}") => {
  let json = JSON.parse(preParser(data));

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
