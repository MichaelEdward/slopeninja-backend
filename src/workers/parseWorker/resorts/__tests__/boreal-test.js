import fs from 'fs';
import {
  parseBorealSnow,
  parseBorealLifts,
  parseBorealTrails,
  parseBorealLiftList,
  parseBorealTrailList,
} from '../boreal';
import { createJSONParser, decodeEntities } from '../../parserFactory';

test('fetches Boreal snow data correctly', async () => {
  const jsonText = fs.readFileSync(`${__dirname}/__fixtures__/boreal-weather.json.input`);

  const resortData = await createJSONParser('snow', parseBorealSnow, decodeEntities)(jsonText);
  expect(resortData).toEqual({
    snow: {
      status: 'closed',
      weatherIcon: 'cloudy',
      temprature: 70,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: 170,
    },
  });
})

test('fetches all null for nonexisting values', async () => {
  const resortData = await createJSONParser('snow', parseBorealSnow, decodeEntities)("{}");
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches all null for no data', async () => {
  const resortData = await createJSONParser('snow', parseBorealSnow, decodeEntities)();
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches Boreal lifts data correctly', async () => {
  const jsonText = fs.readFileSync(`${__dirname}/__fixtures__/boreal-weather.json.input`);
  const resortData = await createJSONParser('lifts', parseBorealLifts, decodeEntities)(jsonText);
  expect(resortData).toEqual({
    lifts: {
      total: 6,
      open: 0,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createJSONParser('lifts', parseBorealLifts, decodeEntities)("{}");
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches Boreal trails data correctly', async () => {
  const jsonText = fs.readFileSync(`${__dirname}/__fixtures__/boreal-weather.json.input`);
  const resortData = await createJSONParser('trails', parseBorealTrails, decodeEntities)(jsonText);
  expect(resortData).toEqual({
    trails: {
      total: 33,
      open: 0,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createJSONParser('trails', parseBorealTrails, decodeEntities)("{}");
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches Boreal lift list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/__fixtures__/boreal-lifts.json.input`);
  const resortData = await createJSONParser('liftList', parseBorealLiftList, decodeEntities)(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createJSONParser('liftList', parseBorealLiftList, decodeEntities)("{}");
  expect(resortData).toMatchObject({ liftList: [] });
});

test('fetches Boreal trail list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/__fixtures__/boreal-trails.json.input`);
  const resortData = await createJSONParser('trailList', parseBorealTrailList, decodeEntities)(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createJSONParser('trailList', parseBorealTrailList, decodeEntities)("{}");
  expect(resortData).toMatchObject({ trailList: [] });
});
