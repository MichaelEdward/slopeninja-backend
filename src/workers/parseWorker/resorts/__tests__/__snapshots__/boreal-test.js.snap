// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches Boreal lift list correctly 1`] = `
Object {
  "liftList": Array [
    Object {
      "category": "Accelerator Zone",
      "name": "Accelerator Express Quad Charilift",
      "status": "closed",
    },
    Object {
      "category": "Beginner Zone",
      "name": "Castle Peak Quad Chairlift",
      "status": "closed",
    },
    Object {
      "category": "Beginner Zone",
      "name": "Nugget Chairlift",
      "status": "closed",
    },
    Object {
      "category": "49er Zone",
      "name": "49er Triple Chairlift",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "name": "Cedar Ridge Triple Chairlift	",
      "status": "closed",
    },
    Object {
      "category": "Flying Dutchman Zone",
      "name": "Flying Dutchman Triple Chairlift",
      "status": "closed",
    },
  ],
}
`;

exports[`fetches Boreal trail list correctly 1`] = `
Object {
  "trailList": Array [
    Object {
      "category": "Accelerator Zone",
      "level": "beginner",
      "name": "Lower Cut",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "beginner",
      "name": "Racer's Return",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "beginner",
      "name": "Sunrise Blvd.",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "beginner",
      "name": "Sunset Blvd.",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "intermediate",
      "name": "Lower Klondike",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "intermediate",
      "name": "His",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "intermediate",
      "name": "Prospector",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "intermediate",
      "name": "Sierra",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "advanced",
      "name": "East Face",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "advanced",
      "name": "Klondike",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "advanced",
      "name": "Race Course",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "advanced",
      "name": "Road Island",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "advanced",
      "name": "Stagecoach",
      "status": "closed",
    },
    Object {
      "category": "Accelerator Zone",
      "level": "advanced",
      "name": "West Bowl",
      "status": "closed",
    },
    Object {
      "category": "Beginner Zone",
      "level": "beginner",
      "name": "Kiss A Bear",
      "status": "closed",
    },
    Object {
      "category": "Beginner Zone",
      "level": "beginner",
      "name": "Rick's Return",
      "status": "closed",
    },
    Object {
      "category": "Beginner Zone",
      "level": "beginner",
      "name": "South 40",
      "status": "closed",
    },
    Object {
      "category": "Beginner Zone",
      "level": "beginner",
      "name": "Walt's Run",
      "status": "closed",
    },
    Object {
      "category": "49er Zone",
      "level": "intermediate",
      "name": "Ptarmigan",
      "status": "closed",
    },
    Object {
      "category": "49er Zone",
      "level": "intermediate",
      "name": "West Face",
      "status": "closed",
    },
    Object {
      "category": "49er Zone",
      "level": "advanced",
      "name": "Chukker",
      "status": "closed",
    },
    Object {
      "category": "49er Zone",
      "level": "advanced",
      "name": "Mogul Rock",
      "status": "closed",
    },
    Object {
      "category": "49er Zone",
      "level": "advanced",
      "name": "Waterfall",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "level": "intermediate",
      "name": "Home Run",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "level": "intermediate",
      "name": "Mariposa",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "level": "intermediate",
      "name": "Tamarack",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "level": "advanced",
      "name": "Hers",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "level": "advanced",
      "name": "Juniper",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "level": "advanced",
      "name": "Ponderosa",
      "status": "closed",
    },
    Object {
      "category": "Cedar Ridge Zone",
      "level": "advanced",
      "name": "Yours",
      "status": "closed",
    },
    Object {
      "category": "Flying Dutchman Zone",
      "level": "intermediate",
      "name": "Central Pacific ",
      "status": "closed",
    },
    Object {
      "category": "Flying Dutchman Zone",
      "level": "intermediate",
      "name": "D. & R. G.",
      "status": "closed",
    },
    Object {
      "category": "Flying Dutchman Zone",
      "level": "intermediate",
      "name": "Westword Ho",
      "status": "closed",
    },
  ],
}
`;
